### PR TITLE
[configure] Add option to use a locally built dotnet/runtime.

### PR DESCRIPTION
**Pull request by rolfbjarne**
* Add a configure option to use a locally built dotnet/runtime.
* Add documentation how to build dotnet/runtime the way we need it built.
* Modify our build to consume the custom dotnet/runtime if so configured.

This is useful when trying to debug the runtime locally, or trying out new
features there are no packages for yet.

### CODE REVIEW

**Review by spouliot — APPROVED** [review]

File: Make.config [hunk 1]
```diff
@@ -512,14 +512,24 @@ DOTNET_FEED_DIR ?= $(DOTNET_DESTDIR)/nuget-feed
 # avoid consuming a lot of disk space (since they're never automatically deleted). The system-dependencies.sh script will install locally as long
 # as there's a TARBALL url.
 
+# If you build dotnet/runtime locally, "*-dev" is the version the built artifactes will use (and which we must reference)
```
Comment (tj-devel709) [inline Make.config:515]:
```suggestion
# If you build dotnet/runtime locally, "*-dev" is the version the built artifacts will use (and which we must reference)
```
I think a small typo

**Review by tj-devel709 — APPROVED** [review]

**Review by stephen-hawley — APPROVED** [review]
Like the documentation

**Comment by mandel-macaque** [comment]
> /Users/builder/azdo/_work/1/s/xamarin-macios/builds/package-download/download-packages.proj : error NU1102: Unable to find package microsoft.netcore.app.ref with version (= 6.0.100-preview.4.21218.6)
> /Users/builder/azdo/_work/1/s/xamarin-macios/builds/package-download/download-packages.proj : error NU1102:   - Found 1347 version(s) in dotnet6 [ Nearest version: 6.0.0-preview.6.21269.9 ]
> /Users/builder/azdo/_work/1/s/xamarin-macios/builds/package-download/download-packages.proj : error NU1102:   - Found 1120 version(s) in dotnet5 [ Nearest version: 6.0.0-alpha.1.20420.3 ]
> /Users/builder/azdo/_work/1/s/xamarin-macios/builds/package-download/download-packages.proj : error NU1102:   - Found 28 version(s) in dotnet-public [ Nearest version: 6.0.0-preview.3.21201.4 ]
> /Users/builder/azdo/_work/1/s/xamarin-macios/builds/package-download/download-packages.proj : error NU1102:   - Found 1 version(s) in Dotnet arcade [ Nearest version: 5.0.0-alpha.1.19618.1 ]
> /Users/builder/azdo/_work/1/s/xamarin-macios/builds/package-download/download-packages.proj : error NU1102:   - Found 0 version(s) in xamarin-impl
> /Users/builder/azdo/_work/1/s/xamarin-macios/builds/package-download/download-packages.proj : error NU1102:   - Found 0 version(s) in macios-dependencies

**Review by mandel-macaque — APPROVED** [review]

File: tests/Directory.Build.targets [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+	<ItemGroup Condition="'$(CUSTOM_DOTNET_VERSION)' != ''">
+		<FrameworkReference Update="Microsoft.NETCore.App" RuntimeFrameworkVersion="$(CUSTOM_DOTNET_VERSION)" />
+		<KnownFrameworkReference Update="Microsoft.NETCore.App" TargetingPackVersion="$(CUSTOM_DOTNET_VERSION)" />
+		<KnownFrameworkReference Update="Microsoft.NETCore.App" RuntimeFrameworkVersion="$(CUSTOM_DOTNET_VERSION)" />
+
+		<FrameworkReference Update="Microsoft.NETCore.App.Mono" RuntimeFrameworkVersion="$(CUSTOM_DOTNET_VERSION)" />
+		<KnownFrameworkReference Update="Microsoft.NETCore.App.Mono" TargetingPackVersion="$(CUSTOM_DOTNET_VERSION)" />
+		<KnownFrameworkReference Update="Microsoft.NETCore.App.Mono" RuntimeFrameworkVersion="$(CUSTOM_DOTNET_VERSION)" />
+	</ItemGroup>
+</Project>
```
Comment (mandel-macaque) [inline tests/Directory.Build.targets:11]: @rolfbjarne 👀

Comment (rolfbjarne) [reply tests/Directory.Build.targets]: Fixed :smile:

File: tests/Directory.Build.targets [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+	<ItemGroup Condition="'$(CUSTOM_DOTNET_VERSION)' != ''">
+		<FrameworkReference Update="Microsoft.NETCore.App" RuntimeFrameworkVersion="$(CUSTOM_DOTNET_VERSION)" />
+		<KnownFrameworkReference Update="Microsoft.NETCore.App" TargetingPackVersion="$(CUSTOM_DOTNET_VERSION)" />
+		<KnownFrameworkReference Update="Microsoft.NETCore.App" RuntimeFrameworkVersion="$(CUSTOM_DOTNET_VERSION)" />
+
+		<FrameworkReference Update="Microsoft.NETCore.App.Mono" RuntimeFrameworkVersion="$(CUSTOM_DOTNET_VERSION)" />
+		<KnownFrameworkReference Update="Microsoft.NETCore.App.Mono" TargetingPackVersion="$(CUSTOM_DOTNET_VERSION)" />
+		<KnownFrameworkReference Update="Microsoft.NETCore.App.Mono" RuntimeFrameworkVersion="$(CUSTOM_DOTNET_VERSION)" />
```
Comment (akoeplinger) [inline tests/Directory.Build.targets:9]: @rolfbjarne curious, why did you need this? we don't produce a package with that name afaik.

Comment (rolfbjarne) [reply tests/Directory.Build.targets:9]: The idea was to pick up the runtime packs from the locally build dotnet/runtime, but it wouldn't surprise me if this isn't necessary (or doesn't do anything useful), it took me a while and some stumbling to figure out how to make the build pick up the locally built runtime packs, and this might have been a shot in the dark that missed completely.

Comment (akoeplinger) [reply tests/Directory.Build.targets:9]: Ok, yeah this shouldn't do anything and can be removed.

Here's how we redirect the runtime pack location to a folder in dotnet/runtime: https://github.com/dotnet/runtime/blob/bdc5d79e9976aeeeabe2a6e38b1e80e33bd01506/eng/targetingpacks.targets#L116-L117
